### ISSUE TITLE
Put operation over an existing object raises a KeyError (-406000 locked data object)

### ISSUE THREAD

**Issue by kjsanger** [irods/python-irodsclient#285]
iRODS server version: 4.2.10
python-irodsclient version: 1.0

**What did you do?:**

Ran `iRODSSession.data_objects.put` to upload a file:
```
INFO:irods.parallel:num_threads = 3 ; (P)artitionSize = 334236926
{"coll": "/testZone/home/irods", "obj": "test.dat"}

real	0m2.108s
user	0m0.461s
sys	0m0.559s
```

Then ran the same operation again to overwrite the file (or raise `OVERWRITE_WITHOUT_FORCE_FLAG`
as I'm not sure how to supply `force` to an overwrite with this API).

**What happened:**

The script crashed with `KeyError: -406000`

**What did you expect:**

The data object to be updated (overwritten) or `OVERWRITE_WITHOUT_FORCE_FLAG` to be
raised. Instead a  `locked data object` error is raised, which the API can't handle because
it's absent from  `irods.exception`.

a) The `locked data object` exception should be known to the API and be present in the
`iRODSExceptionMeta` table. The error should really be `OVERWRITE_WITHOUT_FORCE_FLAG`

or

b) Each API release should advertise which version(s) of iRODS it is compatible with.

The error was 100% reproducible if the file written to already exists. Relevant part of the stack trace:

```
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/manager/data_object_manager.py", line 115, in put
    if not self.parallel_put( local_path, (obj,o), num_threads = num_threads,
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/manager/data_object_manager.py", line 166, in parallel_put
    return parallel.io_main( self.sess, data_or_path_, parallel.Oper.PUT | (parallel.Oper.NONBLOCKING if async_ else 0), file_,
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/parallel.py", line 400, in io_main
    retval = _io_multipart_threaded (Operation, (Data, Io), replica_token, resc_hier, session, fname, total_bytes,
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/parallel.py", line 325, in _io_multipart_threaded
    Io = session.data_objects.open( D.path, Operation.data_object_mode(initial_open = False),
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/manager/data_object_manager.py", line 252, in open
    desc = conn.recv().int_info
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/connection.py", line 105, in recv
    raise get_exception_by_code(msg.int_info)
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/exception.py", line 74, in get_exception_by_code
    return iRODSExceptionMeta.codes[code](message)
KeyError: -406000
```

Edit: found a much better reproducer, removed irrelevant details about timing and iput.

**Comment by d-w-moore** [comment]
@kjsanger Based on the description, I have tried to reproduce with the following script but without success:
```python
#!/usr/bin/env python
# - issue_285.py

from __future__ import print_function
from irods.test.helpers import make_session
import logging
logging.basicConfig(level = logging.INFO)

with make_session() as ses:

  # == create 33 mebibyte file
  with open( '33meb', 'wb') as f: 
    for i in range(33): f.write((b'_'*127+b'\n')*8192)
  print('-- 1st put --')
  ses.data_objects.put( '33meb', '/tempZone/home/rods/33meb' )

  with open( '33meb', 'r+b') as f: 
    f.seek(1024); f.write(b'hidden message')
  print('-- 2nd put --')
  ses.data_objects.put( '33meb', '/tempZone/home/rods/33meb' )
```
My results are:
```
$ ils -L
$ python ~/issue_285.py 
-- 1st put --
INFO:irods.parallel:num_threads = 3 ; (P)artitionSize = 11534337
-- 2nd put --
INFO:irods.parallel:num_threads = 3 ; (P)artitionSize = 11534337
$ iget 33meb /tmp -f
$ diff 33meb /tmp && echo Yes
Yes
$ grep hidden /tmp/33meb 
hidden message_____ [...]
```

Is it possible you could send more detail about the code that is producing the error on your side?

**Comment by kjsanger** [comment]
Taking your script and adjusting the path to my local zone name (image at https://hub.docker.com/r/wsinpg/ub-18.04-irods-4.2.10):

```
#!/usr/bin/env python
# - issue_285.py

from __future__ import print_function
from irods.test.helpers import make_session
import logging
logging.basicConfig(level = logging.INFO)

with make_session() as ses:
  # == create 33 mebibyte file
  with open( '33meb', 'wb') as f:
    for i in range(33): f.write((b'_'*127+b'\n')*8192)
  print('-- 1st put --')
  ses.data_objects.put( '33meb', '/testZone/home/irods/33meb' )

  with open( '33meb', 'r+b') as f:
    f.seek(1024); f.write(b'hidden message')
  print('-- 2nd put --')
  ses.data_objects.put( '33meb', '/testZone/home/irods/33meb' )
```

I see;

```
$ ./bin/issue_285.py 
-- 1st put --
INFO:irods.parallel:num_threads = 3 ; (P)artitionSize = 11534337
Traceback (most recent call last):
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/connection.py", line 103, in recv
    err_msg = iRODSMessage(msg=msg.error).get_main_message(Error).RErrMsg_PI[0].msg
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/message/__init__.py", line 192, in get_main_message
    msg.unpack(ET.fromstring(self.msg))
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/xml/etree/ElementTree.py", line 1347, in XML
    parser.feed(text)
TypeError: a bytes-like object is required, not 'NoneType'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/keith/dev/work/python/pikestaff/./bin/issue_285.py", line 14, in <module>
    ses.data_objects.put( '33meb', '/testZone/home/irods/33meb' )
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/manager/data_object_manager.py", line 115, in put
    if not self.parallel_put( local_path, (obj,o), num_threads = num_threads,
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/manager/data_object_manager.py", line 166, in parallel_put
    return parallel.io_main( self.sess, data_or_path_, parallel.Oper.PUT | (parallel.Oper.NONBLOCKING if async_ else 0), file_,
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/parallel.py", line 400, in io_main
    retval = _io_multipart_threaded (Operation, (Data, Io), replica_token, resc_hier, session, fname, total_bytes,
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/parallel.py", line 325, in _io_multipart_threaded
    Io = session.data_objects.open( D.path, Operation.data_object_mode(initial_open = False),
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/manager/data_object_manager.py", line 252, in open
    desc = conn.recv().int_info
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/connection.py", line 105, in recv
    raise get_exception_by_code(msg.int_info)
  File "/home/keith/.local/miniconda/envs/pikestaff/lib/python3.9/site-packages/irods/exception.py", line 74, in get_exception_by_code
    return iRODSExceptionMeta.codes[code](message)
KeyError: -406000
```

```
$ ilsresc
demoResc:unixfilesystem
replResc:replication
├── unixfs1:unixfilesystem
└── unixfs2:unixfilesystem
testResc:unixfilesystem

$ ils -L
/testZone/home/irods:
  irods             0 replResc;unixfs2      4194304 2021-09-03.09:43 & 33meb
        generic    /var/lib/irods/iRODS/VaultRy/home/irods/33meb
  irods             1 replResc;unixfs1      4194304 2021-09-03.09:43 & 33meb
        generic    /var/lib/irods/iRODS/VaultRx/home/irods/33meb

$ ienv
irods_version - 4.2.10
irods_environment_file - /home/keith/.irods/irods_environment.json
irods_port - 1247
irods_zone_name - testZone
irods_cwd - /testZone/home/irods
irods_user_name - irods
irods_default_resource - replResc
irods_default_hash_scheme - MD5
irods_home - /testZone/home/irods
irods_session_environment_file - /home/keith/.irods/irods_environment.json.9270
irods_host - localhost
```

**Comment by trel** [comment]
I see Keith is targeting `replResc`... Dan, please test/confirm that behavior.

**Comment by d-w-moore** [comment]
Reproduced it on a simple hierarchy, no replResc necessary

**Comment by d-w-moore** [comment]
Not even a resource hierarchy is necessary, actually `parallel_put` is not selecting the correct replica on one or more threads.

**Comment by trel** [comment]
We've determined this is a server issue (irods/irods#5848) and has been fixed and will be included in 4.2.11.

There are no code changes necessary in PRC, but we're going to add tests that capture this Python use case.
